src/Psalm/Internal/Analyzer/Statements/Expression/CastAnalyzer.php

```diff
@@ -300,6 +289,444 @@ public static function analyze(
         return false;
     }
 
+    public static function castIntAttempt(
+        StatementsAnalyzer $statements_analyzer,
+        Context $context,
+        Union $stmt_type,
+        PhpParser\Node\Expr $stmt,
+        bool $explicit_cast = false
+    ): Union {
+        $codebase = $statements_analyzer->getCodebase();
+
+        $risky_cast = [];
+        $invalid_casts = [];
+        $valid_ints = [];
+        $castable_types = [];
+
+        $atomic_types = $stmt_type->getAtomicTypes();
+
+        $parent_nodes = [];
+
+        if ($statements_analyzer->data_flow_graph instanceof VariableUseGraph) {
+            $parent_nodes = $stmt_type->parent_nodes;
+        }
+
+        while ($atomic_types) {
+            $atomic_type = array_pop($atomic_types);
+
+            if ($atomic_type instanceof TInt) {
+                $valid_ints[] = $atomic_type;
+
+                continue;
+            }
+
+            if ($atomic_type instanceof TFloat) {
+                if ($atomic_type instanceof TLiteralFloat) {
+                    $valid_ints[] = new TLiteralInt((int) $atomic_type->value);
+                } else {
+                    $castable_types[] = new TInt();
+                }
+
+                continue;
+            }
+
+            if ($atomic_type instanceof TString) {
+                if ($atomic_type instanceof TLiteralString) {
+                    $valid_ints[] = new TLiteralInt((int) $atomic_type->value);
+                } elseif ($atomic_type instanceof TNumericString) {
+                    $castable_types[] = new TInt();
+                } else {
+                    // any normal string is technically $valid_int[] = new TLiteralInt(0);
+                    // however we cannot be certain that it's not inferred, therefore less strict
+                    $castable_types[] = new TInt();
+                }
+
+                continue;
+            }
+
+            if ($atomic_type instanceof TNull || $atomic_type instanceof TFalse) {
+                $valid_ints[] = new TLiteralInt(0);
+                continue;
+            }
+
+            if ($atomic_type instanceof TBool) {
+                $valid_ints[] = new TLiteralInt(1);
+                continue;
+            }
```

Comment:
Either you meant `TTrue` instead of `TBool`, or this should be `int<0, 1>`. Probably handle `TTrue` first and then `TBool`?

Reply:
> No, this is correct. TFalse is already handled in the condition above (TLiteralInt(0)) therefore bool is only TTrue here.

Actually no. The type `false` is represented by `TFalse`, `true` is represented by `TTrue`, but bool isn't `Union([new TFalse(), new TTrue()])`, it's `TBool`, so you can absolutely have a `TBool` that's neither `TTrue` nor `TFalse`.